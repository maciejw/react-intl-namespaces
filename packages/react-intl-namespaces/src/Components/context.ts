import * as PropTypes from 'prop-types';
import { IntlBackendProvider } from './IntlBackendProvider';

export interface TranslatedMessages {
  [key: string]: string;
}

export namespace IntlBackendContext {
  export type GetMessagesFromNamespace = (
    namespace: string,
    includeNamespace: string[],
  ) => void;

  export type AddMissingMessage = (message: MessageMetadata) => void;

  export type RegisterNamespaceDownloadNotification = (
    callback: (
      resource: { namespace: string; messages: TranslatedMessages },
    ) => void,
  ) => void;

  export type GetIntlProps = () => ReactIntl.IntlProvider.Props;
  export interface MessageMetadata {
    key: string;
    namespace: string;
    defaultMessage?: string;
    description?: string;
  }

  export interface DevConfigProps {
    includeMetadata: boolean;
    showIds: boolean;
  }
  interface IntlBackend extends DevConfigProps {
    getMessagesFromNamespace: GetMessagesFromNamespace;
    getIntlProps: GetIntlProps;
    addMissingMessage: AddMissingMessage;
    registerNamespaceDownloadNotification: RegisterNamespaceDownloadNotification;
  }
  export interface Context {
    intlBackend: IntlBackend;
  }

  const intlBackendShape: PropTypes.ValidationMap<IntlBackend> = {
    getMessagesFromNamespace: PropTypes.func.isRequired,
  };

  const intlContextTypes: React.ValidationMap<Context> = {
    intlBackend: PropTypes.shape(intlBackendShape),
  };

  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}

export namespace IntlNamespaceContext {
  interface IntlNamespace extends IntlBackendContext.DevConfigProps {
    getNameCurrentNamespace(): string;
    missingMessage(message: ReactIntl.FormattedMessage.MessageDescriptor): void;
  }
  export interface Context {
    intlNamespace: IntlNamespace;
  }
  const intlNamespaceShape: PropTypes.ValidationMap<IntlNamespace> = {
    missingMessage: PropTypes.func.isRequired,
  };
  const intlContextTypes: React.ValidationMap<Context> = {
    intlNamespace: PropTypes.shape(intlNamespaceShape),
  };

  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}
