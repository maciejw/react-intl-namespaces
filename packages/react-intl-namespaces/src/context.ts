import * as PropTypes from 'prop-types';
import { ResourceFromNamespace } from './types';

export namespace IntlBackendContext {
  export type GetMessagesFromNamespace = (
    namespaceLoadedNotification: (resource: ResourceFromNamespace) => void,
    namespace: string,
    includeNamespace: string[],
  ) => void;

  export type AddMissingMessage = (message: MessageMetadata) => void;

  export type GetIntlProps = () => ReactIntl.IntlProvider.Props;
  export interface MessageMetadata {
    key: string;
    namespace: string;
    defaultMessage?: string;
    description?: string;
  }

  export interface DevConfigProps {
    includeMetadata: boolean;
    showIds: boolean;
    loggingEnabled: boolean;
  }
  export interface IntlBackend extends DevConfigProps {
    getMessagesFromNamespace: GetMessagesFromNamespace;
    getIntlProps: GetIntlProps;
    addMissingMessage: AddMissingMessage;
  }
  export interface Context {
    intlBackend: IntlBackend;
  }

  const intlBackendShape: PropTypes.ValidationMap<IntlBackend> = {
    getMessagesFromNamespace: PropTypes.func.isRequired,
  };

  const intlContextTypes: React.ValidationMap<Context> = {
    intlBackend: PropTypes.shape(intlBackendShape),
  };
  /**
   * decorator function that defines consumer end of react context
   * @param target component that acquires context
   */
  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  /**
   * decorator function that defines publisher end of react context
   * @param target component that creates context
   */
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}

export namespace IntlNamespaceContext {
  export interface IntlNamespace extends IntlBackendContext.DevConfigProps {
    getNameOfCurrentNamespace(): string;
    missingMessage(message: ReactIntl.FormattedMessage.MessageDescriptor): void;
  }
  export interface Context {
    intlNamespace: IntlNamespace;
  }
  const intlNamespaceShape: PropTypes.ValidationMap<IntlNamespace> = {
    missingMessage: PropTypes.func.isRequired,
  };
  export const intlContextTypes: React.ValidationMap<Context> = {
    intlNamespace: PropTypes.shape(intlNamespaceShape),
  };

  /**
   * decorator function that defines consumer end of react context
   * @param target component that acquires context
   */
  export function Define<P>(target: React.ComponentClass<P>) {
    target.childContextTypes = intlContextTypes;
  }
  /**
   * decorator function that defines publisher end of react context
   * @param target component that creates context
   */
  export function Provide<P>(target: React.ComponentClass<P>) {
    target.contextTypes = { ...target.contextTypes, ...intlContextTypes };
  }
}
