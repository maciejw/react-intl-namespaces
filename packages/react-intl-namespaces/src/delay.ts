export interface Cancelable {
  cancel: () => void;
}

export class CancelablePromise<T> implements Promise<T>, Cancelable {
  public [Symbol.toStringTag]: 'Promise';
  private promise: Promise<T>;
  constructor(
    executor: (
      resolve: (value?: T | PromiseLike<T>) => void,
      reject: (reason?: any) => void,
    ) => void,
    public cancel: () => void,
  ) {
    this.promise = new Promise(executor);
  }

  public then<TResult1 = T, TResult2 = never>(
    onfulfilled?:
      | ((value: T) => TResult1 | PromiseLike<TResult1>)
      | null
      | undefined,
    onrejected?:
      | ((reason: any) => TResult2 | PromiseLike<TResult2>)
      | null
      | undefined,
  ): Promise<TResult1 | TResult2> {
    return this.promise.then(onfulfilled, onrejected);
  }
  public catch<TResult = never>(
    onrejected?:
      | ((reason: any) => TResult | PromiseLike<TResult>)
      | null
      | undefined,
  ): Promise<T | TResult> {
    return this.promise.catch(onrejected);
  }
}

export function delay(timeout: number = 0): Promise<void> & Cancelable {
  let timeoutHandle = 0;

  const p = new CancelablePromise<void>(
    resolve => {
      timeoutHandle = window.setTimeout(resolve, timeout);
    },
    () => window.clearTimeout(timeoutHandle),
  );

  return p;
}
