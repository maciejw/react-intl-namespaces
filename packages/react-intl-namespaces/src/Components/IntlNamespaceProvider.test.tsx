import { mount } from 'enzyme';
import * as React from 'react';
import {
  defineMessages,
  FormattedMessage,
  IntlNamespaceProvider,
} from 'react-intl-namespaces';

import { IntlBackendContext } from '../context';
import { delay } from '../delay';

const messages = defineMessages<'property1'>({
  property1: {
    defaultMessage: 'text1 {someValue}',
    description: 'description1',
    id: 'property1',
  },
});

const values = { someValue: 'value1' };

describe('IntlNamespaceProvider', () => {
  const sut = (
    <IntlNamespaceProvider
      namespace="namespace"
      includeNamespace={['anotherNamespace']}
    >
      <div>
        <FormattedMessage
          id="anotherNamespace:property1"
          defaultMessage="test2 {someValue}"
          values={values}
        />

        <FormattedMessage {...messages.property1} values={values} />
      </div>
    </IntlNamespaceProvider>
  );
  const IntlBackendMock = jest.fn<IntlBackendContext.IntlBackend>(() => ({
    addMissingMessage: jest.fn(),
    getIntlProps: jest.fn(() => ({
      locale: 'en',
    })),
    getMessagesFromNamespace: jest.fn(async notification => {
      await delay();
      notification({
        namespace: 'anotherNamespace',
        resource: {
          property1: 'translated text2 {someValue}',
        },
      });
      notification({
        namespace: 'namespace',
        resource: {
          property1: 'translated text1 {someValue}',
        },
      });
    }),
  }));

  it('should provide translated messages and show translations', async () => {
    const intlBackendMock = new IntlBackendMock({});

    const wrapper = mount(sut, {
      context: {
        intlBackend: intlBackendMock,
      },
    });

    await delay();

    expect(wrapper.html()).toMatchSnapshot();
    expect(intlBackendMock).toMatchSnapshot();
  });

  it('should provide translations and but show message keys', async () => {
    const intlBackendMock = new IntlBackendMock();

    const wrapper = mount(sut, {
      context: {
        intlBackend: intlBackendMock,
      },
    });

    await delay();

    expect(wrapper.html()).toMatchSnapshot();
    expect(intlBackendMock).toMatchSnapshot();
  });
});
