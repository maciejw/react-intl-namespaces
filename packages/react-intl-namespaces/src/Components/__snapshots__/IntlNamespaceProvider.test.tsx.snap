// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`IntlNamespaceProvider should provide translated messages and show translations 1`] = `"<div><span>translated text2 value1</span><span>translated text1 value1</span></div>"`;

exports[`IntlNamespaceProvider should provide translated messages and show translations 2`] = `
Object {
  "addMissingMessage": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "defaultMessage": "test2 {someValue}",
          "description": undefined,
          "key": "property1",
          "namespace": "anotherNamespace",
        },
      ],
      Array [
        Object {
          "defaultMessage": "text1 {someValue}",
          "description": "description1",
          "key": "property1",
          "namespace": "namespace",
        },
      ],
      Array [
        Object {
          "defaultMessage": "text1 {someValue}",
          "description": "description1",
          "key": "property1",
          "namespace": "namespace",
        },
      ],
    ],
  },
  "getIntlProps": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
    ],
  },
  "getMessagesFromNamespace": [MockFunction] {
    "calls": Array [
      Array [
        [Function],
        "namespace",
        Array [
          "anotherNamespace",
        ],
      ],
    ],
  },
}
`;

exports[`IntlNamespaceProvider should provide translations and but show message keys 1`] = `"<div><span>translated text2 value1</span><span>translated text1 value1</span></div>"`;

exports[`IntlNamespaceProvider should provide translations and but show message keys 2`] = `
Object {
  "addMissingMessage": [MockFunction] {
    "calls": Array [
      Array [
        Object {
          "defaultMessage": "test2 {someValue}",
          "description": undefined,
          "key": "property1",
          "namespace": "anotherNamespace",
        },
      ],
      Array [
        Object {
          "defaultMessage": "text1 {someValue}",
          "description": "description1",
          "key": "property1",
          "namespace": "namespace",
        },
      ],
      Array [
        Object {
          "defaultMessage": "text1 {someValue}",
          "description": "description1",
          "key": "property1",
          "namespace": "namespace",
        },
      ],
    ],
  },
  "getIntlProps": [MockFunction] {
    "calls": Array [
      Array [],
      Array [],
      Array [],
    ],
  },
  "getMessagesFromNamespace": [MockFunction] {
    "calls": Array [
      Array [
        [Function],
        "namespace",
        Array [
          "anotherNamespace",
        ],
      ],
    ],
  },
}
`;
